File: shapes.py
from math import pi

class Shape:
    def __init__(self):
        raise TypeError
    def __getPerimetr(self):
        raise TypeError
    def __getArea(self):
        raise TypeError
    perimetr = property(__getPerimetr)
    area = property(__getArea)
    def __str__(self):
        return f"perimetr= {self.perimetr:f} area= {self.area:f}"

class Circle(Shape):
    def __init__(self, r):
        if r <=0:
            raise ValueError("The side can't be negative")
        self.__r = r
    def __getPerimetr(self):
        return self.__r * 2 * pi
    perimetr = property(__getPerimetr)
    def __getArea(self):
        return self.__r**2 * pi
    area = property(__getArea)

    def __str__(self):
        return super().__str__()

class Rectangle(Shape):
    def __init__(self, sides):
        if len(sides) not in [2]:
            raise ValueError(f"{type(self)} must have 2 sides")
        if any(x <= 0 for x in sides):
            raise ValueError("The side can't be negative")
        self.__sides = sides

    def __getPerimetr(self):
        return sum(self.__sides) * 2
    perimetr= property(__getPerimetr)

    def __getArea(self):
        return self.__sides[0] * self.__sides[1]
    area = property(__getArea)

    def __getDiagonal(self):
        return sum(x**2 for x in self.__sides)**0.5
    diagonal = property(__getDiagonal)

    def __str__(self):
        return super().__str__()
    

class Square(Rectangle):
    def __init__(self, side):
        if side <= 0:
            raise ValueError("The side can't be negative")
        super().__init__([side, side])
        self.__side = side
    def __getInRadius(self):
        return self.__side/2
    inRadius = property(__getInRadius)


class Triangle(Shape):
    def __init__(self, sides):
        if len(sides) not in [3]:
            raise ValueError("Triangle must have 3 sides ")
        if any(x <= 0 for x in sides):
            raise ValueError("The side of triangle can't be negative")
        if any(sides[i] + sides[(i + 1)%3] <= sides[(i + 2)%3] for i in range(3)):
            raise ValueError(f"It can't be sides of {type(self)}: {sides}")
        self.__sides = sides
    
    def __getPerimetr(self):
        return sum(self.__sides)
    perimetr= property(__getPerimetr)

    def __getArea(self):
        p = self.perimetr/2
        ans = 1
        for s in self.__sides:
            ans *= (p - s)
        ans *= p
        return ans**0.5
    area = property(__getArea)
    
    def __getInRadius(self):
        return self.area/(self.perimetr/2)
    inRadius = property(__getInRadius)

    def __str__(self):
        return super().__str__()

t = Triangle([3,4,5])
s = Square(5)
r = Rectangle([4,6])
c = Circle(10)
for x in [t,s,r,c]:
    print(x)
for x in [t, s]:
    print("радиус вписанной окружности=", x.inRadius)

File: labs/shapes/shapes.cpp
#include <cmath>
#include <iostream>
#include <math.h>
#include <vector>
using namespace std;
class Shape {
public:
    float virtual get_area()=0;
    float virtual get_perimetr()=0;
    //virtual ~Shape() = default;
};
class Circle : public Shape {
    float r;
public:
    static const float pi;
    Circle(float r):r(r){}

    float get_area() override{
        return pi*r*r;
    }
    float get_perimetr() override {
        return pi*2*r;
    }
};
const float Circle::pi = 3.14;

class Rectangle : public Shape {
    float h;
    float d;
    public:
    Rectangle(float h, float d):h(h),d(d){}
    float get_area() override {
        return d*h;
    }
    float get_perimetr() override {
        return 2*(d+h);
    }
    virtual float get_diagonal() {
        return d*sqrt(2);
    }
};
class Square : public Rectangle {
    public:
    Square(float h):Rectangle(h,h){}
};
int main() {
    vector<Shape*> shapes;
    shapes.push_back(new Circle(10));
    shapes.push_back(new Rectangle(3, 10));
    shapes.push_back(new Square(20));
    float sum_area = 0;
    float sum_perim = 0;
    cout << dynamic_cast<Rectangle *>(shapes[0]);//->get_diagonal() << endl;
    for (auto shape : shapes) {
        sum_area += shape->get_area();
        sum_perim += shape->get_perimetr();
        delete shape;
    }
    cout << "Area: " << sum_area << endl;
    cout << "Perim: " << sum_perim << endl;
}